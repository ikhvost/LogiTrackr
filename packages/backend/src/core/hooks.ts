import { interfaces } from 'inversify'

export enum Hook {
  /** Runs on container boot */
  Bootstrap = 'Hook.Bootstrap',

  /** Runs on container start */
  Start = 'Hook.Start',

  /** Runs on container stop */
  Stop = 'Hook.Stop',
}

export interface HookFn {
  (container: interfaces.Container): Promise<unknown> | unknown

  priority?: number
}

export function withPriority(priority: number, fn: HookFn): HookFn {
  fn.priority = priority

  return fn
}

export async function emit(container: interfaces.Container, hook: Hook, once = true): Promise<interfaces.Container> {
  const hookInvokedKey = `${hook}.Invoked`

  if (once && container.isBound(hookInvokedKey)) {
    return container
  }

  if (container.isBound(hook)) {
    const functions = container.getAll<HookFn>(hook).sort((a, b) => {
      // Use nullish coalescing operator to provide a default value of 0 for priority
      const priorityA = a.priority ?? 0
      const priorityB = b.priority ?? 0

      // Return negative value if b's priority is higher to sort in descending order
      return priorityB - priorityA
    })

    for (const fn of functions) {
      await fn(container)
    }
  }

  if (!container.isBound(hookInvokedKey)) {
    container.bind(hookInvokedKey).toConstantValue(true)
  }

  return container
}
