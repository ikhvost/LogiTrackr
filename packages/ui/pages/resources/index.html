<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Audit Log</title>
    <link href="https://fonts.googleapis.com/css2?family=Poppins:wght@300;400;500;600;700&display=swap"
        rel="stylesheet">
    <link rel="stylesheet" href="./styles.css">
</head>

<body class="text-gray-800 antialiased">
    <div class="max-w-7xl mx-auto py-12 px-4 sm:px-6 lg:px-8">
        <!-- Search Section -->
        <div class="search-container rounded-3xl shadow-lg mb-12 p-10">
            <h1 class="text-4xl font-bold text-white mb-8">Advanced Audit Log</h1>
            <div class="flex flex-col md:flex-row justify-between items-center space-y-4 md:space-y-0 md:space-x-4">
                <input id="searchInput" type="text" placeholder="Search by ID or type..."
                    class="w-full md:w-3/4 p-4 rounded-full border-0 focus:ring-2 focus:ring-indigo-300 transition text-gray-700 text-lg" />
                <button id="searchButton"
                    class="btn-primary w-full md:w-auto px-10 py-4 rounded-full text-white text-lg font-semibold shadow-lg hover:shadow-xl transition-all duration-300 ease-in-out">
                    Search
                </button>
            </div>
        </div>

        <!-- Table Section -->
        <div class="bg-white rounded-2xl shadow-lg overflow-hidden">
            <table class="min-w-full divide-y divide-gray-200">
                <thead class="table-header">
                    <tr>
                        <th class="px-6 py-4 text-left text-sm font-semibold text-gray-600 uppercase tracking-wider">ID
                        </th>
                        <th class="px-6 py-4 text-left text-sm font-semibold text-gray-600 uppercase tracking-wider">
                            Type</th>
                        <th class="px-6 py-4 text-left text-sm font-semibold text-gray-600 uppercase tracking-wider">
                            Last Updated</th>
                        <th class="px-6 py-4 text-left text-sm font-semibold text-gray-600 uppercase tracking-wider">
                            Version</th>
                    </tr>
                </thead>
                <tbody id="resultsBody" class="divide-y divide-gray-200">
                    <!-- Dynamic content will be injected here -->
                </tbody>
            </table>
        </div>

        <!-- Pagination Section -->
        <div class="flex justify-between items-center mt-8">
            <p id="entriesInfo" class="text-sm text-gray-500"></p>
            <div class="flex space-x-2 items-center">
                <button id="prevPage"
                    class="text-gray-600 px-4 py-2 rounded-lg bg-white hover:bg-gray-50 disabled:opacity-50 transition-all shadow-sm"
                    disabled>«</button>
                <div id="pageNumbers" class="text-sm text-gray-600"></div>
                <button id="nextPage"
                    class="text-gray-600 px-4 py-2 rounded-lg bg-white hover:bg-gray-50 disabled:opacity-50 transition-all shadow-sm"
                    disabled>»</button>
            </div>
        </div>
    </div>

    <script type="module" src="scripts/main.ts"></script>
</body>

</html>