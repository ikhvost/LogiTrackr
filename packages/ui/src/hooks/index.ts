export * from './useSearch'
export * from './useRevisions'
