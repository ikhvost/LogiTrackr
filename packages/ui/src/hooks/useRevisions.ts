import { useState, useEffect } from 'react'
import { Metadata, Paginated, Version } from '@logitrackr/contracts'
import { fetcher } from '../utils'

export function useRevisions(id: string, page: number, pageSize: number) {
  const [revisions, setRevisions] = useState<Version[] | null>(null)
  const [metadata, setMetadata] = useState<Metadata>({ totalCount: 0, totalPages: 0, currentPage: 1 })
  const [error, setError] = useState<Error | null>(null)

  useEffect(() => {
    const loadRevisions = async () => {
      setError(null)
      try {
        const response = await fetcher.get<Paginated<Version>>(`/resources/${id}/versions`, {
          page,
          size: pageSize,
        })
        setRevisions(response.data)
        setMetadata(response.metadata)
      } catch (err) {
        console.error('Error fetching revisions:', err)
        setError(err instanceof Error ? err : new Error('An error occurred while fetching data'))
        setRevisions(null)
      }
    }

    loadRevisions().catch(console.error)
  }, [id, page, pageSize])

  return { revisions, metadata, error }
}
