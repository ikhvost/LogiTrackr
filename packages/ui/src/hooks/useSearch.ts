import { useState, useEffect } from 'react'
import { Metadata, Paginated, SearchResponse as Record } from '@logitrackr/contracts'
import { fetcher } from '../utils'

export function useSearch(query: string, page: number, pageSize: number) {
  const [records, setRecords] = useState<Record[] | null>(null)
  const [metadata, setMetadata] = useState<Metadata>({ totalCount: 0, totalPages: 0, currentPage: 1 })
  const [error, setError] = useState<Error | null>(null)

  useEffect(() => {
    const loadData = async () => {
      setError(null)
      try {
        const response = await fetcher.get<Paginated<Record>>('/search', { q: query, page, size: pageSize })
        setRecords(response.data)
        setMetadata(response.metadata)
      } catch (err) {
        console.error('Error fetching data:', err)
        setError(err instanceof Error ? err : new Error('An error occurred while fetching data'))
        setRecords(null)
      }
    }

    loadData().catch(console.error)
  }, [query, page, pageSize])

  return { records, metadata, error }
}
