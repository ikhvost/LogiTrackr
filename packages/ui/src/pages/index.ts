export * from './Resources'
export * from './Versions'
