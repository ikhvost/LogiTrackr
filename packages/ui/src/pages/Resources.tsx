import React, { useState } from 'react'
import { useNavigate } from 'react-router-dom'
import { SearchResponse as Record } from '@saas-versioning/contracts'
import { Table, Pagination, Search } from '../components'
import { useSearch } from '../hooks'

export const Resources = () => {
  const columns = React.useMemo(
    () => [
      {
        header: 'ID',
        accessor: 'id' as keyof Record,
        render: (record: Record) => record.externalId,
      },
      { header: 'Type', accessor: 'type' as keyof Record },
      {
        header: 'Last Updated',
        accessor: 'updatedAt' as keyof Record,
        render: (record: Record) => new Date(record.updatedAt).toLocaleString(),
      },
      {
        header: 'Version',
        accessor: 'lastVersion' as keyof Record,
        render: (record: Record) => record.lastVersion.revision,
      },
    ],
    [],
  )

  const actions = {
    search: (newQuery: string) => {
      setQuery(newQuery)
      setPage(1)
    },
    prevPage: () => {
      if (page > 1) {
        setPage(page - 1)
      }
    },
    nextPage: () => {
      if (page < metadata.totalPages) {
        setPage(page + 1)
      }
    },
    onRowClick: (record: Record) => {
      navigate(`/${record.id}`, { state: { externalId: record.externalId } })
    },
    itemsPerPage: 10,
  }

  const navigate = useNavigate()
  const [query, setQuery] = useState('')
  const [page, setPage] = useState(1)
  const { records, metadata, error } = useSearch(query, page, actions.itemsPerPage)

  if (error) return <div className="text-red-600">Error: {error.message}</div>

  return (
    <>
      <div className="bg-gradient-to-r from-indigo-500 to-purple-500 rounded-3xl shadow-lg mb-12 p-10">
        <h1 className="text-4xl font-bold text-white mb-8">Advanced Audit Log</h1>
        <Search onSearch={actions.search} />
      </div>
      <Table<Record> data={records || []} columns={columns} onRowClick={actions.onRowClick} />
      <Pagination
        metadata={metadata}
        onPrevPage={actions.prevPage}
        onNextPage={actions.nextPage}
        itemsPerPage={actions.itemsPerPage}
      />
    </>
  )
}
