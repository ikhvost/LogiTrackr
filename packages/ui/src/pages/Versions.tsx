import React from 'react'
import { useParams } from 'react-router-dom'
import { Version } from '@logitrackr/contracts'
import { Table, Pagination, Diff, BackButton } from '../components'
import { useRevisions } from '../hooks/useRevisions'

interface Props {
  externalId?: string
}

export const Versions: React.FC<Props> = ({ externalId }) => {
  const columns = React.useMemo(
    () => [
      { header: 'ID', accessor: 'id' as keyof Version },
      {
        header: 'Created At',
        accessor: 'createdAt' as keyof Version,
        render: (revision: Version) => new Date(revision.createdAt).toLocaleString('en-US', { timeZone: 'UTC' }),
      },
      { header: 'Revision', accessor: 'revision' as keyof Version },
    ],
    [],
  )

  const actions = {
    prevPage: () => {
      if (page > 1) {
        setPage(page - 1)
      }
    },
    nextPage: () => {
      if (page < metadata.totalPages) {
        setPage(page + 1)
      }
    },
    itemsPerPage: 10,
  }

  const { id } = useParams<{ id: string }>()
  const [page, setPage] = React.useState(1)
  const { revisions, metadata, error } = useRevisions(id!, page, actions.itemsPerPage)

  const renderExpandedRow = (revision: Version) => {
    const currentIndex = revisions?.findIndex((r) => r.id === revision.id) ?? -1
    const previousRevision = currentIndex < revisions!.length - 1 ? revisions![currentIndex + 1] : null

    return (
      <div className="bg-gray-50 p-6 rounded-lg shadow-inner">
        <h3 className="text-xl font-semibold mb-4">Data Comparison for Revision {revision.revision}</h3>
        {previousRevision ? (
          <Diff oldData={previousRevision.data} currentData={revision.data} />
        ) : (
          <p className="text-gray-600 italic">No previous version to compare.</p>
        )}
      </div>
    )
  }

  if (error) return <div className="text-red-600">Error: {error.message}</div>

  return (
    <div className="container mx-auto px-4 py-8">
      <div className="mb-8 bg-gradient-to-r from-blue-500 to-purple-600 rounded-lg shadow-lg">
        <div className="px-6 py-4 flex items-center justify-between">
          <div>
            <h1 className="text-3xl font-bold text-white mb-2">Resource Revisions</h1>
            {externalId && <p className="text-blue-100">External ID: {externalId}</p>}
          </div>
          <BackButton />
        </div>
      </div>
      <div className="bg-white rounded-lg shadow-md overflow-hidden">
        <Table<Version> data={revisions || []} columns={columns} renderExpandedRow={renderExpandedRow} />
      </div>
      <Pagination
        metadata={metadata}
        onPrevPage={actions.prevPage}
        onNextPage={actions.nextPage}
        itemsPerPage={actions.itemsPerPage}
      />
    </div>
  )
}
