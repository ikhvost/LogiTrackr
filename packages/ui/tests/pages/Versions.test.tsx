import React from 'react'
import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import { describe, test, expect, vi, beforeEach } from 'vitest'
import { useParams, useNavigate, NavigateFunction } from 'react-router-dom'
import { Versions } from '../../src/pages/Versions'
import { useRevisions } from '../../src/hooks/useRevisions'

vi.mock('../../src/hooks/useRevisions')
vi.mock('react-router-dom')

describe('<Versions />', () => {
  const mockMetadata = { totalCount: 2, totalPages: 1, currentPage: 1 }
  const mockRevisions = [
    { id: 'test-id-1', revision: 1, createdAt: '2023-01-01T12:00:00Z', data: {} },
    { id: 'test-id-2', revision: 2, createdAt: '2023-01-02T12:00:00Z', data: {} },
  ]

  beforeEach(() => {
    vi.mocked(useRevisions).mockReturnValue({ revisions: mockRevisions, metadata: mockMetadata, error: null })
    vi.mocked(useParams).mockReturnValue({ id: '123' })
    vi.mocked(useNavigate).mockReturnValue(vi.fn() as unknown as NavigateFunction)
  })

  test('rendering: component renders without crashing', () => {
    const { container } = render(<Versions />)
    expect(container).toMatchSnapshot()
  })

  test('displays correct content', () => {
    render(<Versions externalId="ext123" />)

    expect(screen.getByText('Resource Revisions')).toBeInTheDocument()
    expect(screen.getByText('External ID: ext123')).toBeInTheDocument()
    expect(screen.getByText('test-id-1')).toBeInTheDocument()
    expect(screen.getByText('test-id-2')).toBeInTheDocument()
  })

  test('handles pagination', async () => {
    vi.mocked(useRevisions).mockReturnValue({
      revisions: mockRevisions,
      metadata: { ...mockMetadata, totalPages: 2 },
      error: null,
    })

    render(<Versions />)

    const nextButton = screen.getByText('Â»')
    fireEvent.click(nextButton)

    await waitFor(() => {
      expect(useRevisions).toHaveBeenCalledWith('123', 2, 10)
    })
  })

  test('displays error message when error occurs', () => {
    vi.mocked(useRevisions).mockReturnValue({ revisions: null, metadata: mockMetadata, error: new Error('Test error') })

    render(<Versions />)

    expect(screen.getByText('Error: Test error')).toBeInTheDocument()
  })

  test('renders expanded row on row click', async () => {
    render(<Versions />)

    const firstRow = screen.getByText('1').closest('tr')
    fireEvent.click(firstRow!)

    await waitFor(() => {
      expect(screen.getByText('Data Comparison for Revision 1')).toBeInTheDocument()
    })
  })
})
