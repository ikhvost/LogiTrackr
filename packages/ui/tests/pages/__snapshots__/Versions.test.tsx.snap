// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Versions /> > rendering: component renders without crashing 1`] = `
<div>
  <div
    class="container mx-auto px-4 py-8"
  >
    <div
      class="mb-8 bg-gradient-to-r from-blue-500 to-purple-600 rounded-lg shadow-lg"
    >
      <div
        class="px-6 py-4 flex items-center justify-between"
      >
        <div>
          <h1
            class="text-3xl font-bold text-white mb-2"
          >
            Resource Revisions
          </h1>
        </div>
        <button
          class="flex items-center px-4 py-2 bg-blue-500 text-white rounded hover:bg-blue-600 transition-colors"
        >
          <svg
            class="h-5 w-5 mr-2"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 19l-7-7m0 0l7-7m-7 7h18"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
          Back
        </button>
      </div>
    </div>
    <div
      class="bg-white rounded-lg shadow-md overflow-hidden"
    >
      <table
        class="min-w-full divide-y divide-gray-200"
      >
        <thead
          class="bg-gray-50"
        >
          <tr>
            <th
              class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
              scope="col"
            >
              ID
            </th>
            <th
              class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
              scope="col"
            >
              Created At
            </th>
            <th
              class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
              scope="col"
            >
              Revision
            </th>
          </tr>
        </thead>
        <tbody
          class="bg-white divide-y divide-gray-200"
        >
          <tr
            class="cursor-pointer hover:bg-gray-100 "
          >
            <td
              class="px-6 py-4 whitespace-nowrap"
            >
              test-id-1
            </td>
            <td
              class="px-6 py-4 whitespace-nowrap"
            >
              1/1/2023, 12:00:00 PM
            </td>
            <td
              class="px-6 py-4 whitespace-nowrap"
            >
              1
            </td>
          </tr>
          <tr
            class="cursor-pointer hover:bg-gray-100 "
          >
            <td
              class="px-6 py-4 whitespace-nowrap"
            >
              test-id-2
            </td>
            <td
              class="px-6 py-4 whitespace-nowrap"
            >
              1/2/2023, 12:00:00 PM
            </td>
            <td
              class="px-6 py-4 whitespace-nowrap"
            >
              2
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="flex justify-between items-center mt-8"
    >
      <p
        class="text-sm text-gray-500"
      >
        Showing 
        2
         of 
        2
         entries
      </p>
      <div
        class="flex space-x-2 items-center"
      >
        <button
          class="text-gray-600 px-4 py-2 rounded-lg bg-white hover:bg-gray-50 disabled:opacity-50 transition-all shadow-sm"
          disabled=""
        >
          «
        </button>
        <div
          class="text-sm text-gray-600"
        >
          Page 
          1
           of 
          1
        </div>
        <button
          class="text-gray-600 px-4 py-2 rounded-lg bg-white hover:bg-gray-50 disabled:opacity-50 transition-all shadow-sm"
          disabled=""
        >
          »
        </button>
      </div>
    </div>
  </div>
</div>
`;
